src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,160 @@
+#include "XrdCephBulkAioRead.hh"
+
+bulkAioRead::bulkAioRead(librados::IoCtx *ct, logfunc_pointer logwrapper, std::string filename, size_t block_sz) {
+  context = ct;
+  block_size = block_sz;
+  file_name = filename;
+  log_func = logwrapper;
+};
+
+bulkAioRead::~bulkAioRead() {
+  clear();
+};
+
+void bulkAioRead::clear() {
+  for (ReadOpData i: buffers){
+    delete std::get<0>(i);
+    delete std::get<2>(i);
+  }
+  buffers.clear();
+  for (auto const& tup: operations) {
+    librados::ObjectReadOperation *op = std::get<0>(tup.second);
+    librados::AioCompletion *cmpl = std::get<1>(tup.second);
+    cmpl->release();
+    delete op;
+  }
+  operations.clear();
+};
+
+int bulkAioRead::addRequest(std::string objname, char* out_buf, size_t size, off64_t offset) {
+  ceph::bufferlist *bl;
+  int *retval;
+  librados::ObjectReadOperation *rop;
+
+  try {
+    bl = new ceph::bufferlist();
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate buffer for read (%lu, %lu)", offset, size);
+    return -ENOMEM;
+  }
+
+  try {
+    retval = new int;
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate int for read's retval (%lu, %lu)", offset, size);
+    delete bl;
+    return -ENOMEM;
+  }
+
+  auto tup = operations.find(objname);
+  if (operations.end() == tup) {
+    try {
+      rop = new librados::ObjectReadOperation;
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not allocate int for read object op (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      return -ENOMEM;
+    }
+
+    librados::AioCompletion *cmpl = librados::Rados::aio_create_completion();
+    if (0 == cmpl) {
+      log_func((char*)"Can not create completion for read (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      delete rop;
+      return -ENOMEM;
+    }
+    operations.insert( {objname, std::make_pair(rop, cmpl)} );
+  } else {
+    rop = std::get<0>(tup->second);
+  }
+  buffers.push_back( std::make_tuple(bl, out_buf, retval) );
+
+  rop->read(offset, size, bl, retval);
```

Comment:
What about return value?

Reply:
It is void function:
https://github.com/ceph/ceph/blob/main/src/librados/librados_cxx.cc#L218